[SPARK-33427][SQL] Add subexpression elimination for interpreted expression evaluation

viirya opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This patch proposes to add subexpression elimination for interpreted expression evaluation. Interpreted expression evaluation is used when codegen was not able to work, for example complex schema.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Currently we only do subexpression elimination for codegen. For some reasons, we may need to run interpreted expression evaluation. For example, codegen fails to compile and fallbacks to interpreted mode, or complex input/output schema of expressions. It is commonly seen for complex schema from expressions that is possibly caused by the query optimizer too, e.g. SPARK-32945.

We should also support subexpression elimination for interpreted evaluation. That could reduce performance difference when Spark fallbacks from codegen to interpreted expression evaluation, and improve Spark usability.

#### Benchmark

Update `SubExprEliminationBenchmark`:

Before:

```
OpenJDK 64-Bit Server VM 1.8.0_265-b01 on Mac OS X 10.15.6
 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 from_json as subExpr:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 -------------------------------------------------------------------------------------------------------------------------
subexpressionElimination on, codegen off           24707          25688         903          0.0   247068775.9       1.0X
```

After:
```
OpenJDK 64-Bit Server VM 1.8.0_265-b01 on Mac OS X 10.15.6
 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 from_json as subExpr:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 -------------------------------------------------------------------------------------------------------------------------
subexpressionElimination on, codegen off            2360           2435          87          0.0    23604320.7      11.2X
```

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Unit test. Benchmark manually.

viirya committed:
Add subexpression elimination for interpreted expression evaluation.

SparkQA commented:
**[Test build #130948 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130948/testReport)** for PR 30341 at commit [`c7fdae5`](https://github.com/apache/spark/commit/c7fdae569209374a5bfcf038dfc1995935a180da).

viirya reviewed (commented):
The main code change should be ready for review. I marked it as WIP because I'd like to add some tests later.

SparkQA commented:
**[Test build #130948 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130948/testReport)** for PR 30341 at commit [`c7fdae5`](https://github.com/apache/spark/commit/c7fdae569209374a5bfcf038dfc1995935a180da).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class EvaluationRunTime `
  * `case class ExpressionProxy(child: Expression, runtime: EvaluationRunTime) extends Expression `
  * `case class ResultProxy(result: Any)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130948/
Test FAILed.

viirya committed:
Catch cache exception and throw original exception.

SparkQA commented:
**[Test build #130949 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130949/testReport)** for PR 30341 at commit [`33ac8b4`](https://github.com/apache/spark/commit/33ac8b4d4fffb231054d16a80ab6366f30683c2d).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35554/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35554/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35554/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35555/

SparkQA commented:
**[Test build #130949 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130949/testReport)** for PR 30341 at commit [`33ac8b4`](https://github.com/apache/spark/commit/33ac8b4d4fffb231054d16a80ab6366f30683c2d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130949/
Test FAILed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35555/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35555/
Test PASSed.

viirya committed:
Use invalidateAll instead of cleanUp.

SparkQA commented:
**[Test build #130955 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130955/testReport)** for PR 30341 at commit [`6bab83c`](https://github.com/apache/spark/commit/6bab83c6f247910d421aa02eada0d5fdf4610120).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35561/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35561/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35561/
Test PASSed.

viirya committed:
Add tests.

SparkQA commented:
**[Test build #130965 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130965/testReport)** for PR 30341 at commit [`ddd3a96`](https://github.com/apache/spark/commit/ddd3a96db1e22bd0b8c8a5b4c351479a63e1e907).

viirya commented:
GitHub Actions were passed actually.

viirya commented:
cc @cloud-fan @dongjoon-hyun @maropu @HyukjinKwon

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35571/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35571/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35571/
Test PASSed.

dongjoon-hyun commented:
Thank you for pinging me, @viirya .

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRunTime.scala:
```diff
@@ -0,0 +1,88 @@
+ * This class wraps `ExpressionProxy` around given expressions. The `ExpressionProxy`
+ * intercepts expression evaluation and loads from the cache first.
+ */
+class EvaluationRunTime {
```
`Runtime` instead of `RunTime`.
For example, Java has `RuntimeException`. Spark follows it and uses `RunTime` for `time` only.

dongjoon-hyun replied:
Please update the file name too.
`EvaluationRunTime.scala` -> `EvaluationRuntime.scala`

viirya replied:
Oh, good to know. I will change it.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+}
+
+/**
+ * A simple wrapper for holding `Any` in the cache of `EvaluationRunTime`.
```
ditto. `RunTime` -> `Runtime`

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+  override def eval(input: InternalRow = null): Any = try {
+    runtime.cache.get(this).result
+  } catch {
+    // Cache.get() may wrap the original exception. See the following URL
```
Quick question: Is this only `14.0` specific? If we change the latest Guava, is this different?

viirya replied:
Seems the same: https://guava.dev/releases/21.0/api/docs/com/google/common/cache/Cache.html#get-K-java.util.concurrent.Callable-

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+    // Cache.get() may wrap the original exception. See the following URL
+    // http://google.github.io/guava/releases/14.0/api/docs/com/google/common/cache/
+    //   Cache.html#get(K,%20java.util.concurrent.Callable)
+    case e@(_: UncheckedExecutionException | _: ExecutionError) =>
```
nit. maybe spacing? `e@(` -> `e @ (`?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -539,6 +539,15 @@ object SQLConf {
+      .doc("The maximum entries of the cache used for interpreted subexpression elimination.")
+      .version("3.1.0")
+      .intConf
+      .checkValue(maxEntries => maxEntries >= 0, "The maximum must not be negative")
```
What happens `maxEntries = 0`?

dongjoon-hyun replied:
Shall we use a simple form like `.checkValue(_ >= 0)`?

viirya replied:
`maxEntries = 0` basically means no caching.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -539,6 +539,15 @@ object SQLConf {
+      .version("3.1.0")
+      .intConf
+      .checkValue(maxEntries => maxEntries >= 0, "The maximum must not be negative")
+      .createWithDefault(100)
```
`100` is enough?

viirya replied:
This default value is copied from code generator's cache setting. I guess this should be enough.

SparkQA commented:
**[Test build #130955 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130955/testReport)** for PR 30341 at commit [`6bab83c`](https://github.com/apache/spark/commit/6bab83c6f247910d421aa02eada0d5fdf4610120).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130955/
Test PASSed.

viirya committed:
For review comments.

SparkQA commented:
**[Test build #130973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130973/testReport)** for PR 30341 at commit [`e8449ec`](https://github.com/apache/spark/commit/e8449ec6a550973aeb66c6deb2bf98a547b934d5).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35579/
Test FAILed.

SparkQA commented:
**[Test build #130965 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130965/testReport)** for PR 30341 at commit [`ddd3a96`](https://github.com/apache/spark/commit/ddd3a96db1e22bd0b8c8a5b4c351479a63e1e907).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #130973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130973/testReport)** for PR 30341 at commit [`e8449ec`](https://github.com/apache/spark/commit/e8449ec6a550973aeb66c6deb2bf98a547b934d5).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130973/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130965/
Test FAILed.

viirya commented:
retest this please

SparkQA commented:
**[Test build #131012 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131012/testReport)** for PR 30341 at commit [`e8449ec`](https://github.com/apache/spark/commit/e8449ec6a550973aeb66c6deb2bf98a547b934d5).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35618/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35618/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35618/
Test PASSed.

SparkQA commented:
**[Test build #131012 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131012/testReport)** for PR 30341 at commit [`e8449ec`](https://github.com/apache/spark/commit/e8449ec6a550973aeb66c6deb2bf98a547b934d5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131012/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  override def eval(input: InternalRow = null): Any = try {
+    runtime.cache.get(this).result
```
does cache lookup include expression canonilization?

viirya replied:
No, here the semantically equal expressions are linked to same `ExpressionProxy`.

cloud-fan replied:
Can we assign unique IDs(per `EvaluationRuntime`) to `ExpressionProxy`? then the lookup can be more efficient.

viirya replied:
Good idea.

viirya replied:
Oh, but in order to look up using unique ID of `ExpressionProxy`, we need to keep a map (id -> `ExpressionProxy`) inside `EvaluationRuntime`, so we know to call which proxy's `proxyEval` in `CacheLoader.load` function.

viirya replied:
I think it should be fine but want to make sure I get your point correctly.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRuntime.scala:
```diff
@@ -0,0 +1,88 @@
+
+/**
+ * This class helps subexpression elimination for interpreted evaluation
+ * in `InterpretedUnsafeProjection`. It maintains an evaluation cache.
```
We will support interpreted filter in the future, right? maybe `in` -> `such as`

viirya replied:
Yes.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRuntime.scala:
```diff
@@ -0,0 +1,88 @@
+class EvaluationRuntime {
+
+  val cache: LoadingCache[ExpressionProxy, ResultProxy] = CacheBuilder.newBuilder()
+    .maximumSize(SQLConf.get.subexpressionEliminationCacheMaxEntries)
```
for low-level class, maybe it's better to have a parameter for cache capacity, instead of accessing the active conf here. It also makes the UT easier as this class don't rely on sql conf.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRuntime.scala:
```diff
@@ -0,0 +1,88 @@
+
+    expressions.foreach(equivalentExpressions.addExprTree(_))
+
+    var proxyMap = Map.empty[Expression, ExpressionProxy]
```
shall we use a mutable map here?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRuntime.scala:
```diff
@@ -0,0 +1,88 @@
+        // `ExpressionProxy`. But we cannot use `transformUp` because we want to use
+        // subexpressions at higher level. So we `transformDown` until finding first
+        // subexpression.
+        var transformed = false
```
this is hacky. Can we do a manual tree traversal using `mapChildren`?

viirya replied:
Oh, good idea, I like it.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRuntime.scala:
```diff
@@ -0,0 +1,88 @@
+        // subexpressions at higher level. So we `transformDown` until finding first
+        // subexpression.
+        var transformed = false
+        expr.transform {
```
what if the single expr already have repeated sub-exprs? e.g. `SELECT (a+1) + (a+1)`. The code here seems can only replace one expression.

viirya replied:
Good point! I think I missed this point. I will add a test too.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+  final override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode =
+    throw new UnsupportedOperationException(s"Cannot generate code for expression: $this")
+
+  def proxyEval(input: InternalRow = null): Any = {
```
nit: this doesn't need a default value.

viirya replied:
Removed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+}
+
+/**
+ * A simple wrapper for holding `Any` in the cache of `EvaluationRuntime`.
```
why do we need this wrapper?

viirya replied:
Seems `CacheLoader` doesn't accept the second type parameter to be `Any`.

cloud-fan commented:
The idea looks good!  left several comments

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedUnsafeProjection.scala:
```diff
@@ -63,17 +72,21 @@ class InterpretedUnsafeProjection(expressions: Array[Expression]) extends Unsafe
     var i = 0
     while (i < numFields) {
-      values(i) = expressions(i).eval(row)
+      values(i) = proxyExpressions(i).eval(row)
```
Any reason not to apply this into `InterpretedMutableProjection`, too?

viirya replied:
It could be applied to `InterpretedMutableProjection` too. We can add it gradually like filter.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRuntime.scala:
```diff
@@ -0,0 +1,88 @@
+ * This class wraps `ExpressionProxy` around given expressions. The `ExpressionProxy`
+ * intercepts expression evaluation and loads from the cache first.
+ */
+class EvaluationRuntime {
```
This seems to be a specific class for subexpression elimination, but I feel the name looks too general. How about renaming it more concretely? e.g., `SubExprElimination[Eval]Runtime` or something?

viirya replied:
Renamed as `SubExprEvaluationRuntime` now.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+    // Cache.get() may wrap the original exception. See the following URL
+    // http://google.github.io/guava/releases/14.0/api/docs/com/google/common/cache/
+    //   Cache.html#get(K,%20java.util.concurrent.Callable)
+    case e @ (_: UncheckedExecutionException | _: ExecutionError) =>
```
How about moving this exception handling into the `ExpressionRuntime` then making `cache` private just like `CodeGenerator.cache`? https://github.com/apache/spark/blob/f80fe213bd4c5e065d5723816c42302a532be75c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala#L1351-L1359

viirya replied:
Moved.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionProxy.scala:
```diff
@@ -0,0 +1,57 @@
+ * A proxy for an catalyst `Expression`. Given a runtime object `EvaluationRuntime`, when this
+ * is asked to evaluate, it will load the evaluation cache in the runtime first.
+ */
+case class ExpressionProxy(child: Expression, runtime: EvaluationRuntime) extends Expression {
```
Seems like this is just a thin wrapper for eliminating common subexprs in `ExpressionRuntime`, so how about putting this class and `ExpressionRuntime` in a single file?

viirya replied:
As a single file `SubExprEvaluationRuntime.scala` now.

maropu commented:
Nice feature!

maropu commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/EvaluationRuntimeSuite.scala:
```diff
@@ -0,0 +1,79 @@
+
+import org.apache.spark.SparkFunSuite
+
+class EvaluationRuntimeSuite extends SparkFunSuite {
```
How about adding some end-2-end tests for this feature in `SQLQueryTestSuite` with `--CONFIG_DIM spark.sql.codegen.wholeStage=false,spark.sql.codegen.factoryMode=NO_CODEGEN` ?

viirya replied:
Okay, let me see how to add some tests for this.

viirya committed:
For review comments.

SparkQA commented:
**[Test build #131088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131088/testReport)** for PR 30341 at commit [`4780b65`](https://github.com/apache/spark/commit/4780b653f079c2fd4c950d43758587875d89e41a).

viirya commented:
I addressed most of comments except for https://github.com/apache/spark/pull/30341#discussion_r522822589 and https://github.com/apache/spark/pull/30341#discussion_r522904967.

For https://github.com/apache/spark/pull/30341#discussion_r522822589, I will wait for confirmation.

For https://github.com/apache/spark/pull/30341#discussion_r522904967, I will add some more tests later.

viirya commented:
Hmm, actually for https://github.com/apache/spark/pull/30341#discussion_r522904967, that is to add some tests in `SQLQueryTestSuite` with `--CONFIG_DIM spark.sql.codegen.wholeStage=false,spark.sql.codegen.factoryMode=NO_CODEGEN`, I think it should be added as a separate PR and merge the test PR first? Then we can run against with it to check if the result is the same. @maropu

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35691/

maropu commented:
>  I think it should be added as a separate PR and merge the test PR first? Then we can run against with it to check if the result is the same. @maropu

Yea, it looks fine. I'll review this again tonight and thanks for the update.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35691/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35691/
Test PASSed.

SparkQA commented:
**[Test build #131088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131088/testReport)** for PR 30341 at commit [`4780b65`](https://github.com/apache/spark/commit/4780b653f079c2fd4c950d43758587875d89e41a).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131088/
Test FAILed.

viirya commented:
retest this please

SparkQA commented:
**[Test build #131089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131089/testReport)** for PR 30341 at commit [`4780b65`](https://github.com/apache/spark/commit/4780b653f079c2fd4c950d43758587875d89e41a).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35692/

SparkQA commented:
**[Test build #131089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131089/testReport)** for PR 30341 at commit [`4780b65`](https://github.com/apache/spark/commit/4780b653f079c2fd4c950d43758587875d89e41a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131089/
Test FAILed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35692/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35692/
Test PASSed.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedUnsafeProjection.scala:
```diff
@@ -33,6 +34,15 @@ import org.apache.spark.unsafe.Platform
 class InterpretedUnsafeProjection(expressions: Array[Expression]) extends UnsafeProjection {
   import InterpretedUnsafeProjection._
 
+  private[this] val subExprElimination = SQLConf.get.subexpressionEliminationEnabled
```
nit: `subExprElimination` -> `subExprEliminationEnabled`?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedUnsafeProjection.scala:
```diff
@@ -33,6 +34,15 @@ import org.apache.spark.unsafe.Platform
+  private[this] val subExprElimination = SQLConf.get.subexpressionEliminationEnabled
+  private[this] lazy val runtime =
+    new SubExprEvaluationRuntime(SQLConf.get.subexpressionEliminationCacheMaxEntries)
+  private[this] val proxyExpressions = if (subExprElimination) {
```
nit: In case of `subExprElimination=false` or no common subexprs, this variable is not related to a "proxy", so how about just calling it `exprs`? (`proxyExpressions` -> `exprs`)

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SubExprEvaluationRuntime.scala:
```diff
@@ -0,0 +1,131 @@
+      case e if proxyMap.contains(e) =>
+        proxyMap(e)
+      case _ =>
+        expr.mapChildren(replaceWithProxy(_, proxyMap))
```
nit: `proxyMap.getOrElse(expr, expr.mapChildren(replaceWithProxy(_, proxyMap)))`?

maropu reviewed (commented):
I left the minor comments and the other parts looks fine.

viirya committed:
For review comment and add one test.

viirya commented:
Thanks @maropu

SparkQA commented:
**[Test build #131094 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131094/testReport)** for PR 30341 at commit [`47bae35`](https://github.com/apache/spark/commit/47bae358690c550a0259796b7bed137d1d38f28e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35697/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35697/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35697/
Test PASSed.

dongjoon-hyun commented:
It would be great if we can have a benchmark suite (from the code in the PR description), @viirya .

viirya commented:
> It would be great if we can have a benchmark suite (from the code in the PR description), @viirya .

Yes, @dongjoon-hyun, I remember it. :) I will add a benchmark suite and the tests @maropu suggested in separate PRs.

SparkQA commented:
**[Test build #131094 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131094/testReport)** for PR 30341 at commit [`47bae35`](https://github.com/apache/spark/commit/47bae358690c550a0259796b7bed137d1d38f28e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131094/
Test PASSed.

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you so much, @viirya .

Please let us know if you have another comments, @maropu and @cloud-fan .

Also, #30379 is a benchmark PR for this. ~After merging this, we can merge that, too.~

viirya commented:
Thanks @dongjoon-hyun.

Let's wait for @cloud-fan's comment for https://github.com/apache/spark/pull/30341#discussion_r522822589 after this weekend.

dongjoon-hyun commented:
#30379 is merged. Could you rebased to the master and regenerate the result, @viirya .

Sure, that's the exact reason I asked their opinion once more. 😄 
> Let's wait for @cloud-fan's comment for #30341 (comment) after this weekend.

viirya commented:
@dongjoon-hyun Thanks! I will update this tonight or tomorrow.

dongjoon-hyun commented:
No need to hurry~ Take your time. I just want to give you feedback swiftly from my side.

viirya committed:
Merge remote-tracking branch 'upstream/master' into SPARK-33427

viirya committed:
Update benchmark results.

SparkQA commented:
**[Test build #131106 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131106/testReport)** for PR 30341 at commit [`77168fe`](https://github.com/apache/spark/commit/77168fe2dd687113ae1b9a2f086982861445ecda).

SparkQA commented:
**[Test build #131106 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131106/testReport)** for PR 30341 at commit [`77168fe`](https://github.com/apache/spark/commit/77168fe2dd687113ae1b9a2f086982861445ecda).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131106/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35709/

viirya commented:
retest this please

SparkQA commented:
**[Test build #131107 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131107/testReport)** for PR 30341 at commit [`77168fe`](https://github.com/apache/spark/commit/77168fe2dd687113ae1b9a2f086982861445ecda).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35709/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35709/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35710/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35710/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35710/
Test PASSed.

maropu commented:
> Please let us know if you have another comments, @maropu and @cloud-fan .

Yea, I don't have any more comment, so I'll leave this to @cloud-fan .

SparkQA commented:
**[Test build #131107 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131107/testReport)** for PR 30341 at commit [`77168fe`](https://github.com/apache/spark/commit/77168fe2dd687113ae1b9a2f086982861445ecda).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131107/
Test PASSed.

dongjoon-hyun commented on sql/core/benchmarks/SubExprEliminationBenchmark-jdk11-results.txt:
```diff
@@ -7,9 +7,9 @@ OpenJDK 64-Bit Server VM 11.0.9+11 on Mac OS X 10.15.6
+subexpressionElimination off, codegen on           25932          26908         916          0.0   259320042.3       1.0X
+subexpressionElimination off, codegen off          26085          26159          65          0.0   260848905.0       1.0X
+subexpressionElimination on, codegen on             2860           2939          72          0.0    28603312.9       9.1X
+subexpressionElimination on, codegen off            2517           2617          93          0.0    25165157.7      10.3X
```
Thank you for this additional `10.3x`.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SubExprEvaluationRuntime.scala:
```diff
@@ -0,0 +1,126 @@
+  private def replaceWithProxy(
+      expr: Expression,
+      proxyMap: Map[Expression, ExpressionProxy]): Expression = {
+    proxyMap.getOrElse(expr, expr.mapChildren(replaceWithProxy(_, proxyMap)))
```
This is a top-down traverse, which means we will recursively replace expr with proxy even for `ExpressionProxy`. Is it expected?

viirya replied:
Once we replace one expression with `ExpressionProxy`, we stop traversing down. We only traverse down to children if cannot find current expression in `proxyMap`. Is this for your question?

cloud-fan replied:
Ah sorry I misread the code. You are right.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SubExprEvaluationRuntime.scala:
```diff
@@ -0,0 +1,126 @@
+
+    expressions.foreach(equivalentExpressions.addExprTree(_))
+
+    val proxyMap = mutable.Map.empty[Expression, ExpressionProxy]
```
is it OK to use a simple map here? Two expressions may not equal to each other even if they semantically equal.

viirya replied:
For semantically equal exprs, we put a pair of expr -> proxy into the map and note the proxy is the same. So later we traverse down into expressions, we look at the map. We don't do semantically comparing when looking at this map.

cloud-fan replied:
Ah I get it now. Seems we can use `IdentityHashMap` to be more explicit.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubExprEvaluationRuntimeSuite.scala:
```diff
@@ -0,0 +1,99 @@
+    // ( (one * two) * (one * two) )
+    assert(proxys.size == 2)
+    val expected = ExpressionProxy(mul2, runtime)
+    assert(proxys.head == expected)
```
should this be `proxys.forall(_ == expected)`?

viirya replied:
yeah, you're right.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubExprEvaluationRuntimeSuite.scala:
```diff
@@ -0,0 +1,99 @@
+    })
+    assert(proxys.isEmpty)
+  }
+}
```
can we test attributes?
```
val attr1 = AttributeReference("a", ...)
val attr2 = attr1.withName("A")
```
To make sure 2 semantically-equal attributes can be optimized.

viirya replied:
`EquivalentExpressions` skips for `LeafExpression`. So attributes won't be counted for subexpression.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SubExprEvaluationRuntime.scala:
```diff
@@ -0,0 +1,126 @@
+ * A proxy for an catalyst `Expression`. Given a runtime object `SubExprEvaluationRuntime`,
+ * when this is asked to evaluate, it will load from the evaluation cache in the runtime first.
+ */
+case class ExpressionProxy(
```
for https://github.com/apache/spark/pull/30341#discussion_r523299091

We can define `ExpressionProxy` as `case class ExpressionProxy(child: Expression, id: Int, ...)` and override `equals`/`hashCode`.

viirya replied:
Okay, but in this case we cannot use `IdentityHashMap` anymore as `IdentityHashMap` doesn't use `equals` to compare keys.

viirya replied:
Ah, nvm. I figured it out.

viirya committed:
For review comment.

SparkQA commented:
**[Test build #131184 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131184/testReport)** for PR 30341 at commit [`db115d6`](https://github.com/apache/spark/commit/db115d6d850942eca5dc6fac80896b1038561e51).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35786/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35786/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35786/
Test FAILed.

SparkQA commented:
**[Test build #131184 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131184/testReport)** for PR 30341 at commit [`db115d6`](https://github.com/apache/spark/commit/db115d6d850942eca5dc6fac80896b1038561e51).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131184/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

dongjoon-hyun commented:
Thank you, @viirya , @maropu , @cloud-fan !

viirya commented:
Thanks @dongjoon-hyun @maropu @cloud-fan

HyukjinKwon commented:
Sorry for a late comment. +1, nice.

viirya commented:
Thanks @HyukjinKwon!
